File: composey/compiler/generator_gcp.py
"""GCP Terraform JSON generator.

Converts GcpResources to Terraform JSON format for the Google provider.
"""

import json
from typing import Any

from composey.models.environment import GcpEnvironment
from composey.models.gcp import GcpResources


def generate(resources: GcpResources, env: GcpEnvironment) -> str:
    """Generate Terraform JSON for GCP resources.

    Args:
        resources: The GCP resources to generate
        env: The GCP environment configuration

    Returns:
        JSON string containing Terraform configuration
    """
    terraform = {
        "terraform": {
            "required_providers": {
                "google": {
                    "source": "hashicorp/google",
                    "version": "~> 5.0",
                },
                "docker": {
                    "source": "kreuzwerker/docker",
                    "version": "~> 3.0",
                },
                "random": {
                    "source": "hashicorp/random",
                    "version": "~> 3.6",
                },
            },
        },
        "provider": {
            "google": {
                "project": env.project_id,
                "region": env.region,
            },
            "docker": {},
            "random": {},
        },
        "resource": _build_resources(resources),
    }

    # Add custom endpoint if specified (for testing)
    if env.gcp_endpoint:
        terraform["provider"]["google"]["credentials"] = env.gcp_endpoint

    return json.dumps(terraform, indent=2)


def _build_resources(resources: GcpResources) -> dict[str, Any]:
    """Build the resource section of Terraform JSON."""
    result: dict[str, Any] = {}

    # Cloud Run services
    if resources.google_cloud_run_service:
        result["google_cloud_run_service"] = {
            k: _clean_model(v) for k, v in resources.google_cloud_run_service.items()
        }

    if resources.google_cloud_run_service_iam_member:
        result["google_cloud_run_service_iam_member"] = {
            k: _clean_model(v)
            for k, v in resources.google_cloud_run_service_iam_member.items()
        }

    # Cloud SQL
    if resources.google_sql_database_instance:
        result["google_sql_database_instance"] = {
            k: _clean_model(v)
            for k, v in resources.google_sql_database_instance.items()
        }

    if resources.google_sql_database:
        result["google_sql_database"] = {
            k: _clean_model(v) for k, v in resources.google_sql_database.items()
        }

    # Memorystore (Redis)
    if resources.google_redis_instance:
        result["google_redis_instance"] = {
            k: _clean_model(v) for k, v in resources.google_redis_instance.items()
        }

    # Cloud Storage
    if resources.google_storage_bucket:
        result["google_storage_bucket"] = {
            k: _clean_model(v) for k, v in resources.google_storage_bucket.items()
        }

    if resources.google_storage_bucket_iam_member:
        result["google_storage_bucket_iam_member"] = {
            k: _clean_model(v)
            for k, v in resources.google_storage_bucket_iam_member.items()
        }

    # VPC Connector
    if resources.google_vpc_access_connector:
        result["google_vpc_access_connector"] = {
            k: _clean_model(v) for k, v in resources.google_vpc_access_connector.items()
        }

    # Load balancing
    if resources.google_compute_global_address:
        result["google_compute_global_address"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_global_address.items()
        }

    if resources.google_compute_managed_ssl_certificate:
        result["google_compute_managed_ssl_certificate"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_managed_ssl_certificate.items()
        }

    if resources.google_compute_region_network_endpoint_group:
        result["google_compute_region_network_endpoint_group"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_region_network_endpoint_group.items()
        }

    if resources.google_compute_backend_service:
        result["google_compute_backend_service"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_backend_service.items()
        }

    if resources.google_compute_url_map:
        result["google_compute_url_map"] = {
            k: _clean_model(v) for k, v in resources.google_compute_url_map.items()
        }

    if resources.google_compute_target_https_proxy:
        result["google_compute_target_https_proxy"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_target_https_proxy.items()
        }

    if resources.google_compute_forwarding_rule:
        result["google_compute_forwarding_rule"] = {
            k: _clean_model(v)
            for k, v in resources.google_compute_forwarding_rule.items()
        }

    # Secret Manager
    if resources.google_secret_manager_secret:
        result["google_secret_manager_secret"] = {
            k: _clean_model(v)
            for k, v in resources.google_secret_manager_secret.items()
        }

    if resources.google_secret_manager_secret_version:
        result["google_secret_manager_secret_version"] = {
            k: _clean_model(v)
            for k, v in resources.google_secret_manager_secret_version.items()
        }

    # Docker resources
    if resources.docker_image:
        result["docker_image"] = resources.docker_image

    if resources.docker_registry_image:
        result["docker_registry_image"] = resources.docker_registry_image

    # Random resources
    if resources.random_password:
        result["random_password"] = {
            k: _clean_model(v) for k, v in resources.random_password.items()
        }

    return result


def _clean_model(obj: Any) -> Any:
    """Clean a Pydantic model for Terraform JSON output."""
    if hasattr(obj, "model_dump"):
        data = obj.model_dump(exclude_none=True)
    elif hasattr(obj, "dict"):
        data = obj.dict(exclude_none=True)
    else:
        return obj

    return data

File: composey-go/internal/compiler/generator_gcp.go
package compiler

import (
	"github.com/gecburton/composey/internal/models"
)

// gcpResourceOrder mirrors _build_resources' if-block order in
// generator_gcp.py exactly. Like Azure's generator, GCP's json.dumps has
// no sort_keys=True, so this order is load-bearing for anyone who wants
// byte-identical output against a live Python run -- no golden file
// exists to check it against here, but the same PyDumpsIndent-based
// approach used for Azure is used again regardless, since it is known to
// be the only correct one for a non-sort_keys json.dumps call.
var gcpResourceOrder = []string{
	"google_cloud_run_service",
	"google_cloud_run_service_iam_member",
	"google_sql_database_instance",
	"google_sql_database",
	"google_redis_instance",
	"google_storage_bucket",
	"google_storage_bucket_iam_member",
	"google_vpc_access_connector",
	"google_compute_global_address",
	"google_compute_managed_ssl_certificate",
	"google_compute_region_network_endpoint_group",
	"google_compute_backend_service",
	"google_compute_url_map",
	"google_compute_target_https_proxy",
	"google_compute_forwarding_rule",
	"google_secret_manager_secret",
	"google_secret_manager_secret_version",
	"docker_image",
	"docker_registry_image",
	"random_password",
}

func gcpResourceBlocks(resources *models.GcpResources) PyOrdered {
	nonEmpty := map[string]any{
		"google_cloud_run_service":                     resources.CloudRunService,
		"google_cloud_run_service_iam_member":          resources.CloudRunServiceIamMember,
		"google_sql_database_instance":                 resources.SqlDatabaseInstance,
		"google_sql_database":                          resources.SqlDatabase,
		"google_redis_instance":                        resources.RedisInstance,
		"google_storage_bucket":                        resources.StorageBucket,
		"google_storage_bucket_iam_member":             resources.StorageBucketIamMember,
		"google_vpc_access_connector":                  resources.VpcAccessConnector,
		"google_compute_global_address":                resources.ComputeGlobalAddress,
		"google_compute_managed_ssl_certificate":       resources.ComputeManagedSslCertificate,
		"google_compute_region_network_endpoint_group": resources.ComputeRegionNetworkEndpointGroup,
		"google_compute_backend_service":               resources.ComputeBackendService,
		"google_compute_url_map":                       resources.ComputeUrlMap,
		"google_compute_target_https_proxy":            resources.ComputeTargetHttpsProxy,
		"google_compute_forwarding_rule":               resources.ComputeForwardingRule,
		"google_secret_manager_secret":                 resources.SecretManagerSecret,
		"google_secret_manager_secret_version":         resources.SecretManagerSecretVersion,
		"docker_image":                                 resources.DockerImage,
		"docker_registry_image":                        resources.DockerRegistryImage,
		"random_password":                              resources.RandomPassword,
	}

	result := PyOrdered{}
	for _, resourceType := range gcpResourceOrder {
		value := nonEmpty[resourceType]
		if !isNonEmptyMapGcp(value) {
			continue
		}
		result = append(result, p(resourceType, structToPyOrdered(value)))
	}
	return result
}

func isNonEmptyMapGcp(v any) bool {
	switch m := v.(type) {
	case map[string]models.CloudRunService:
		return len(m) > 0
	case map[string]models.CloudRunServiceIamMember:
		return len(m) > 0
	case map[string]models.CloudSqlInstance:
		return len(m) > 0
	case map[string]models.CloudSqlDatabase:
		return len(m) > 0
	case map[string]models.RedisInstance:
		return len(m) > 0
	case map[string]models.StorageBucket:
		return len(m) > 0
	case map[string]models.StorageBucketIamMember:
		return len(m) > 0
	case map[string]models.VpcConnector:
		return len(m) > 0
	case map[string]models.GlobalAddress:
		return len(m) > 0
	case map[string]models.ComputeManagedSslCertificate:
		return len(m) > 0
	case map[string]models.ComputeRegionNetworkEndpointGroup:
		return len(m) > 0
	case map[string]models.ComputeBackendService:
		return len(m) > 0
	case map[string]models.ComputeUrlMap:
		return len(m) > 0
	case map[string]models.ComputeTargetHttpsProxy:
		return len(m) > 0
	case map[string]models.ComputeForwardingRule:
		return len(m) > 0
	case map[string]models.SecretManagerSecret:
		return len(m) > 0
	case map[string]models.SecretManagerSecretVersion:
		return len(m) > 0
	case map[string]models.DockerImage:
		return len(m) > 0
	case map[string]models.DockerRegistryImage:
		return len(m) > 0
	case map[string]any:
		return len(m) > 0
	default:
		return false
	}
}

// GenerateGcp renders a Terraform JSON manifest for the given GCP
// resources and environment, mirroring generator_gcp.py's generate().
func GenerateGcp(resources *models.GcpResources, env *models.GcpEnvironment) (string, error) {
	requiredProviders := PyOrdered{
		p("google", PyOrdered{p("source", "hashicorp/google"), p("version", "~> 5.0")}),
		p("docker", PyOrdered{p("source", "kreuzwerker/docker"), p("version", "~> 3.0")}),
		p("random", PyOrdered{p("source", "hashicorp/random"), p("version", "~> 3.6")}),
	}

	googleProvider := PyOrdered{
		p("project", env.ProjectID),
		p("region", env.Region),
	}
	if env.GcpEndpoint != nil && *env.GcpEndpoint != "" {
		googleProvider = append(googleProvider, p("credentials", *env.GcpEndpoint))
	}

	provider := PyOrdered{
		p("google", googleProvider),
		p("docker", PyOrdered{}),
		p("random", PyOrdered{}),
	}

	terraformDoc := PyOrdered{
		p("terraform", PyOrdered{p("required_providers", requiredProviders)}),
		p("provider", provider),
		p("resource", gcpResourceBlocks(resources)),
	}

	return PyDumpsIndent(terraformDoc, 2), nil
}
